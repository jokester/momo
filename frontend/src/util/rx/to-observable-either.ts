import { Either, left, right, Right } from 'fp-ts/lib/Either';
import { Observable, from } from 'rxjs';
import { filter, map, materialize } from 'rxjs/operators';

type StreamValueFromPromise<V, R = unknown> = null | Either<V, R>;

/**
 * @param {PromiseLike<Value>} p
 * @returns {Observable<StreamValueFromPromise<Value, Rejection>>} a observable that emits a {@link Either} and complete
 */
export function toObservableEither<Value, Rejection = unknown>(
  p: PromiseLike<Value>,
): Observable<StreamValueFromPromise<Value, Rejection>> {
  return from(p).pipe(
    materialize(),
    map((notification) =>
      notification.do(
        (v) => right(v),
        (e) => left(e),
      ),
    ),
    filter((_) => _),
  );
}
