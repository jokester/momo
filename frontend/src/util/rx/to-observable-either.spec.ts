import { left, right } from 'fp-ts/lib/Either';
import { toObservableEither } from './to-observable-either';
import { of } from 'rxjs';
import { mergeMap, switchMap, take, toArray, startWith } from 'rxjs/operators';

describe('toObservableEither', () => {
  it('converts promise to observable', async () => {
    const o = toObservableEither(Promise.resolve(1));
    expect(await o.toPromise()).toEqual(right(1));
  });

  it('transforms (stream of promise) to (stream of promise result)', async () => {
    const o = of(1, 2, 3, 4, 5).pipe(
      mergeMap((v) => toObservableEither(v % 2 ? Promise.resolve(v) : Promise.reject(v))),
      take(20),
      toArray(),
    );

    expect(await o.toPromise()).toMatchSnapshot();
  });
});
