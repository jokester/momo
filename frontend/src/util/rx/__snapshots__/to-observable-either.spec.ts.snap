// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toObservableEither transforms (stream of promise) to (stream of promise result) 1`] = `
Array [
  Object {
    "_tag": "Right",
    "right": 1,
  },
  Object {
    "_tag": "Left",
    "left": 2,
  },
  Object {
    "_tag": "Right",
    "right": 3,
  },
  Object {
    "_tag": "Left",
    "left": 4,
  },
  Object {
    "_tag": "Right",
    "right": 5,
  },
]
`;
